src/hotspot/share/prims/agentList.cpp

```diff
@@ -0,0 +1,652 @@
+    }
+  }
+};
+
+class AgentThreadEventMark : public AgentEventMark {
+ private:
+  jobject _jthread;
+ public:
+  AgentThreadEventMark(JavaThread* thread) : AgentEventMark(thread) {
+    _jthread = JNIHandles::make_local(thread, thread->threadObj());
+  };
+  jthread jni_thread() { return (jthread)_jthread; }
+};
+
+static void unload_and_delete(Agent* agent, void* library) {
+  assert(agent != nullptr, "invariant");
+  if (!agent->is_static_lib()) {
+    os::dll_unload(library);
+  }
+  delete agent;
+}
+
+// type for the Agent_OnAttach entry point
+extern "C" {
+  typedef jint(JNICALL* OnAttachEntry_t)(JavaVM*, char*, void*);
+}
+
+// Implementation for loading an agent dynamically during runtime, by invoking Agent_OnAttach.
+jint AgentList::load_agent(const char* agent_name, const char* absParam,
+                                  const char* options, outputStream* st) {
+  assert(JvmtiEnvBase::get_phase() == JVMTI_PHASE_LIVE, "not in live phase!");
+  const char* on_attach_symbols[] = AGENT_ONATTACH_SYMBOLS;
+  const size_t num_symbol_entries = ARRAY_SIZE(on_attach_symbols);
+
+  // The abs parameter should be "true" or "false"
+  const bool is_absolute_path = (absParam != nullptr) && (strcmp(absParam, "true") == 0);
+  // Initially marked as invalid. It will be set to valid if we can find the agent
+  Agent* agent = new Agent(agent_name, options, is_absolute_path);
+  agent->_is_dynamic = true;
+
+  // Check for statically linked in agent. If not found then if the path is
+  // absolute we attempt to load the library. Otherwise we try to load it from the standard dll directory.
+  void* library = load_agent_from_executable(agent, &on_attach_symbols[0], num_symbol_entries);
+  if (library == nullptr) {
+    library = agent->is_absolute_path() ? load_agent_from_absolute_path(agent, /* vm exit on error */ false) :
+                                          load_agent_from_relative_path(agent, /* vm exit on error */ false);
+  }
+  if (library != nullptr) {
+    agent->set_os_lib_path(&buffer[0]);
+    agent->set_os_lib(library);
+    agent->set_valid();
+  } else {
+    st->print_cr("%s was not loaded.", agent_name);
+    if (*ebuf != '\0') {
+      st->print_cr("%s", ebuf);
+    }
+    return JNI_ERR;
+  }
+
+  assert(library != nullptr, "invariant");
+  assert(agent->is_valid(), "invariant");
+
+  // The library was loaded so we attempt to lookup and invoke the Agent_OnAttach function
+  OnAttachEntry_t on_attach_entry = CAST_TO_FN_PTR(OnAttachEntry_t,
+                                                   os::find_agent_function(agent, false, &on_attach_symbols[0], num_symbol_entries));
+
+  if (on_attach_entry == nullptr) {
+    // Agent_OnAttach missing - unload library
+    unload_and_delete(agent, library);
+    st->print_cr("%s is not available in %s", on_attach_symbols[0], agent->name());
+    return JNI_ERR;
+  }
+
+  // Invoke the Agent_OnAttach function
+  JavaThread* THREAD = JavaThread::current(); // For exception macros.
```

Comment:
Nit: just use `current` rather than `THREAD` and don't use the exception macros.

Reply:
Ported as is but good point, will update.

Reply:
Updated - cheers.